staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1864,3 +1986,523 @@ func TestForgetWatcher(t *testing.T) {
+
+		expectedWatchResourceVersion int
+	}{
+		// +-----------------+---------------------+-----------------------+
+		// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+		// +=================+=====================+=======================+
+		// | Unset           | true/false          | nil/true/false        |
+		// +-----------------+---------------------+-----------------------+
+		{
+			name: "RV=unset, allowWatchBookmarks=true, sendInitialEvents=nil, expectedRV=0",
+			opts: listOptions(true, nil, ""),
+		},
+		{
+			name:                         "RV=unset, allowWatchBookmarks=true, sendInitialEvents=true, expectedRV=100(storage)",
+			opts:                         listOptions(true, pointer.Bool(true), ""),
+			expectedWatchResourceVersion: 100,
+		},
+		{
+			name: "RV=unset, allowWatchBookmarks=true, sendInitialEvents=false, expectedRV=0",
+			opts: listOptions(true, pointer.Bool(false), ""),
+		},
+		{
+			name: "RV=unset, allowWatchBookmarks=false, sendInitialEvents=nil, expectedRV=0",
+			opts: listOptions(false, nil, ""),
+		},
+		{
+			name:                         "RV=unset, allowWatchBookmarks=false, sendInitialEvents=true, expectedRV=100(storage), legacy",
+			opts:                         listOptions(false, pointer.Bool(true), ""),
+			expectedWatchResourceVersion: 100,
+		},
+		{
+			name: "RV=unset, allowWatchBookmarks=false, sendInitialEvents=false, expectedRV=0",
+			opts: listOptions(false, pointer.Bool(false), ""),
+		},
+		// +-----------------+---------------------+-----------------------+
+		// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+		// +=================+=====================+=======================+
+		// | 0               | true/false          | nil/true/false        |
+		// +-----------------+---------------------+-----------------------+
+		{
+			name: "RV=0, allowWatchBookmarks=true, sendInitialEvents=nil, expectedRV=0",
+			opts: listOptions(true, nil, "0"),
+		},
+		{
+			name: "RV=0, allowWatchBookmarks=true, sendInitialEvents=true, expectedRV=0",
+			opts: listOptions(true, pointer.Bool(true), "0"),
+		},
+		{
+			name: "RV=0, allowWatchBookmarks=true, sendInitialEvents=false, expectedRV=0",
+			opts: listOptions(true, pointer.Bool(false), "0"),
+		},
+		{
+			name: "RV=0, allowWatchBookmarks=false, sendInitialEvents=nil, expectedRV=0",
+			opts: listOptions(false, nil, "0"),
+		},
+		{
+			name: "RV=0, allowWatchBookmarks=false, sendInitialEvents=true, expectedRV=0",
+			opts: listOptions(false, pointer.Bool(true), "0"),
+		},
+		{
+			name: "RV=0, allowWatchBookmarks=false, sendInitialEvents=false, expectedRV=0",
+			opts: listOptions(false, pointer.Bool(false), "0"),
+		},
+		// +-----------------+---------------------+-----------------------+
+		// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+		// +=================+=====================+=======================+
+		// | 105             | true/false          | nil/true/false        |
```

Comment:
This test is slightly broken because the usual case is that you get RV that is lagging behind the etcd version, not being too new.

So I would rather like this test to be switched to rv=95 (something behind the storage).